invitations: Add a new setting `Who can create multiuse invite link`.

Ujjawal3 opened the pull request:
<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Ujjawal3 commented on zerver/models.py:
```diff
@@ -721,6 +725,9 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
 
+    # Names for system group
+    ADMINISTRATORS_GROUP_NAME = "@role:administrators"
+
```
I think it would be fine if we move all entries of system_groups names from `UserGroup` model to `Realm` model as a follow up to this PR.

timabbott replied:
I think an even better plan might be to move them to a central object or dataclass, rather than hanging off a model at all, but agreed this is a good follow-up.

Ujjawal3 commented on zerver/models.py:
```diff
@@ -2106,6 +2123,8 @@ def email_address_is_realm_public(self) -> bool:
 
     def has_permission(self, policy_name: str) -> bool:
+        from zerver.lib.user_groups import is_user_in_group
+
```
I was getting a circular dependency error if it was imported at the beginning of module.

timabbott replied:
This is probably fine but deserves a comment. Later we might refactor this function away from living in `models.py` to some `zerver/lib/` space of its own to resolve that.

Ujjawal3 commented on web/src/settings_data.ts:
```diff
@@ -141,6 +142,16 @@ export function user_can_invite_others_to_realm(): boolean {
+        page_params.realm_create_multiuse_invite_group_id,
+        page_params.user_id,
+    );
+}
```
I initially wrote the function like this 
```
export function user_can_create_muliuse_invite(): boolean {
    return user_groups.is_user_in_group(
        page_params.realm_create_multiuse_invite_group_id,
        people.my_current_user_id()
    );
}
```
But `People` module can not be imported because of circular dependancy.

Ujjawal3 commented:
@zulipbot add "buddy review"

Ujjawal3 commented:
@HemantUmre12 please review this PR.

HemantUmre12 reviewed (commented):
Great work, @Ujjawal3 ! I have added a couple of comments regarding refactoring and small changes. The commits related to cache, flush and import look good to me based on your discussions, but we still need confirmation from others.

Here are a few more points and suggestions:

- Please mention the related issue in the PR description.
- Shouldn't we rename `Who can invite users to this organization` as per #15159?
- The option is not disabled in SETTINGS / ORGANIZATION PERMISSIONS for `administrators`.

There's one more thing: if `Who can invite users to this organization` is set to `Admin` and
 `Who can create invite links` is set to `Admin, moderators, and members`, I can't see "invite users" in the settings menu as a `member`, which shouldn't be the case since both settings are independent.

What I'm proposing is:

- If the user is not part of the configured groups for both settings, hide the "invite users" option.
- If the user is part of any of the settings' configured groups, show "invite users" and disable the option with a tooltip for the setting that the user is not a part of.

You can further discuss this in CZO.

Ujjawal3 commented on web/src/invite.js:
```diff
@@ -351,6 +352,15 @@ function open_invite_user_modal(e) {
+            $("#email_invite_radio_container").addClass("disabled_setting_tooltip");
+
+            $("#generate_multiuse_invite_radio").prop("checked", true);
+            $("#generate_multiuse_invite_radio").trigger("change");
```
I added this code after adding all event listener so that we can avoid some duplicate code by using `$("#generate_multiuse_invite_radio").trigger("change")`

Ujjawal3 commented on zerver/openapi/zulip.yaml:
```diff
@@ -14142,6 +14153,17 @@ paths:
 
                           **Changes**: New in Zulip 4.0 (feature level 51).
 
+                          [who-can-send-invitations]: /help/restrict-account-creation#change-who-can-send-invitations
```
I don't know why git is changing this line rather than the later one.

Ujjawal3 commented on zerver/views/realm.py:
```diff
@@ -279,6 +284,30 @@ def update_realm(
                 data[k] = v
 
+    for setting_name, permissions_configuration in Realm.REALM_PERMISSION_GROUP_SETTINGS.items():
+        request_settings_dict = locals()
```
I liked the suggested approach. So, updated the code accordingly.

Ujjawal3 commented:
@HemantUmre12 please have a look to this PR. I have done the following changes since last review :

- Addressed all the comments.
- Changed the default group to `Moderators` from `Administrators` user_group.
- Changes in the migration setting value of `create_multiuse_invite_group` to preserve the existing permissions as mentioned in issue description.

HemantUmre12 reviewed (commented):
@Ujjawal3, suggested some minor changes. Here are some additional points:

- In some instances, you have directly used the `UserGroup` name. Can we avoid that?
- Please update the API documentation (zulip.yaml) regarding the `create_multiuse_invite_group` in the `/events` responses.

For the live update-
- Instead of updating each 'invite users' link state separately, we can use the `invite-user-link` class, which is shared by all. Something like,
````
$(".invite-user-link").hide();
````
We also have to add live update code for opened 'invite users to Zulip' menu-
- Close the menu if the new configuration doesn't allow it.
- Update the state of 'Generate invite link' and 'Send an email' based on the new configuration. Alternatively, we can choose to simply close the menu.

Ujjawal3 commented:
@HemantUmre12 I have updated the PR and addressed all your comments. 

Regarding using `$(".invite-user-link").hide();` for live-updates I don't think it is better than the current approach because:
1. `$(".invite-user-link").hide();` works only when all the instances are rendered. So there can be some cases when only some instances are rendered and can lead to bugs.
2. We will have to deal with gear_menu separately otherwise we will get the following UI
![Screenshot from 2023-06-19 19-41-11](https://github.com/zulip/zulip/assets/99073049/52ed6c41-b4b8-4572-b72a-1225f80ae69a)

Regarding live-updating invite users menu,I think its a rare situation, So I am not sure whether we want this because we are not doing such minor updates in some other places too. So its better to not do this till someone else reviews the PR.

Ujjawal3 commented on zerver/migrations/0454_set_default_value_for_create_multiuse_invite_group.py:
```diff
@@ -0,0 +1,49 @@
+
+    for realm in Realm.objects.all():
+        if realm.create_multiuse_invite_group is None:
+            if realm.invite_to_realm_policy == realm.POLICY_NOBODY:
```
`rrealm.POLICY_NOBODY` is not `None` so I made it `realm.invite_to_realm_policy is realm.POLICY_NOBODY:`

sahil839 commented on zerver/models.py:
```diff
@@ -2139,6 +2164,12 @@ def has_permission(self, policy_name: str) -> bool:
+        if policy_name in Realm.REALM_PERMISSION_GROUP_SETTING_ID_TO_GROUP_SETTING:
+            allowed_user_group = getattr(
+                self.realm, Realm.REALM_PERMISSION_GROUP_SETTING_ID_TO_GROUP_SETTING[policy_name]
+            )
```
We can have a separate function for group-based settings but I am not sure. We can wait for more feedback.

sahil839 replied:
But this may be fine since we would convert other settings to group based settings eventually.

sahil839 commented on zerver/views/invite.py:
```diff
@@ -197,7 +209,13 @@ def generate_multiuse_invite_backend(
+        raise JsonableError(_("Insufficient permission"))
+
     check_if_owner_required(invite_as, user_profile)
+    check_if_administrator_required(invite_as, user_profile)
```
I think it would be better to combine both these functions for owner and admin and name it accordingly.

sahil839 commented on zerver/models.py:
```diff
@@ -771,6 +779,20 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
+    )
+
+    REALM_PERMISSION_GROUP_SETTING_ID_TO_GROUP_SETTING = dict(
+        create_multiuse_invite_group_id="create_multiuse_invite_group"
```
I get how this is used, but we can directly add "_id" to the setting name. I feel it is unnecessary, but we can wait for more feedback.

timabbott replied:
I'm not opposed to this existing; I guess most places that currently reference `REALM_PERMISSION_GROUP_SETTING_ID_TO_GROUP_SETTING` feel like they should access `REALM_PERMISSION_GROUP_SETTINGS` instead, but maybe that doesn't work because of the `_id` suffix. 

I think the right larger model would be to have `GroupPermissionSetting` have an `id_field_name` field.

sahil839 replied:
Most of the places we use the `REALM_PERMISSION_GROUP_SETTING_ID_TO_GROUP_SETTING` object in a loop. So in those places we can just loop over `REALM_PERMISSION_GROUP_SETTINGS` and add "_id" to the string like we do for stream settings currently.

And a couple of other places where we use this differently is in `has_permission` code in `models.py` and in `views/realm.py` to check whether a setting is a group-based settings to add its new value in `req_group_setting_vars` object.

We can actually directly use `create_multiuse_invite_group` instead of `create_multiuse_invite_group_id` in `has_permission` as we eventually pass user group object in `is_user_in_group`. While this dict is actually helpful to use in the `views/realm.py` code.

sahil839 replied:
But yeah, I agree that having `id_field_name` would be a better plan.

Ujjawal3 replied:
Hmm, keeping this dict is helpful in `views/realm.py`. If we will change the parameters to not have "_id" than we can remove this dict. But in that case I think we should also make changes in name of `page_params` and `events` because our current logic in frontend for changing the settings is based on the assumption that name of the setting received in `page_params` is same as name of the parameter used for changing the setting. Otherwise we need to figure out some approach to distinguish between role based and group based setting in frontend and  may also need to maintain a dict similar to `REALM_PERMISSION_GROUP_SETTING_ID_TO_GROUP_SETTING` in frontend.

Ujjawal3 commented:
@sahil839 I have updated the PR and addressed your comments. So, you can also review backend part now.

sahil839 commented on web/src/settings_invites.js:
```diff
@@ -289,7 +289,10 @@ export function update_invite_users_setting_tip() {
     update_invite_users_setting_tip();
-    if (!settings_data.user_can_invite_others_to_realm()) {
+    if (
+        !settings_data.user_can_invite_others_to_realm() &&
```
Would also be good to rename `user_can_invite_others_to_realm` to something like `user_can_send_email_invite_to_others`.

Ujjawal3 replied:
Should I just rename this function or all the variables related to this setting?

sahil839 replied:
I think both because having `can_invite_others_to_realm` as `False`, feels user is not allowed to create multiuse link as well. Would probably be better to do this change in a separate prep commit.

sahil839 replied:
Ohh, sorry for the confusion here. But I did not meant to rename the database field and the API fields sent through `/events`, `/register`,etc. endpoints. I only meant the variables used internally. It would be better to get feedback from others before renaming the database fields and the fields sent in API responses.

Ujjawal3 replied:
I have updated the PR. Please have a look.

sahil839 commented on web/styles/settings.css:
```diff
@@ -480,8 +480,14 @@ input[type="checkbox"] {
+        margin-top: 10px;
+    }
+
+    .dropdown-list-widget div:first-of-type {
```
A better fix here would to make the selector for setting margin-top to 10px more selective probably by using `input-group` class.

sahil839 commented on zerver/tests/test_realm.py:
```diff
@@ -1207,11 +1215,55 @@ def do_test_realm_update_api(self, name: str) -> None:
+                )
+                or user_group.name == default_group_name
+            ):
+                continue
```
I think we should also test the cases where a user tries to set the setting to these groups and check that these raise an error.

sahil839 reviewed (commented):
Thanks for updating the PR. Posted some more comments.

Also, Alex gave some feedback on the commit messages for initial commits that I took from this PR for #26070. You might also want to update the commit messages in this PR as well.

Ujjawal3 commented:
@sahil839 I have updated the PR as per your comments. Please have a look.

sahil839 commented on zerver/migrations/0455_set_default_value_for_create_multiuse_invite_group.py:
```diff
@@ -0,0 +1,55 @@
+
+    for realm in Realm.objects.all():
+        if realm.create_multiuse_invite_group is None:
+            if realm.invite_to_realm_policy is Realm.POLICY_NOBODY:
```
Would be better to use `==` for comparing values here. We generally use `is` for checking if  value is `None` or not.

sahil839 commented on zerver/actions/realm_settings.py:
```diff
@@ -104,6 +105,43 @@ def do_set_realm_property(
         update_users_in_full_members_system_group(realm, acting_user=acting_user)
 
 
+@transaction.atomic(savepoint=False)
```
This should have `durable=True` and not `savepoint=False` since we do not call this function inside another atomic transaction as of now.

sahil839 reviewed (commented):
Posted a couple of comments, which I missed in the last review. 

Also, the commit message for `invites: Use a single function to check if owner or admin is required.` commit mentions there were two separate functions which is incorrect.

sahil839 reviewed (commented):
Posted one minor comment. Otherwise LGTM. I have also tested it manually and it works fine. 
@timabbott I think you can have a look at this now. There are some unresolved comments above.

timabbott commented on zerver/lib/server_initialization.py:
```diff
@@ -23,20 +24,21 @@ def server_initialized() -> bool:
-    )
-    RealmUserDefault.objects.create(realm=realm)
-    create_system_user_groups_for_realm(realm)
+    with transaction.atomic(durable=True):
```
What's the thinking behind doing this here rather than around all of `create_internal_realm`? The latter certainly seems more correct to me, since then either we have a fully initialized server or a server that's not initialized at all.

timabbott commented on zerver/views/invite.py:
```diff
@@ -29,13 +29,28 @@
 
-def check_if_owner_required(invited_as: int, user_profile: UserProfile) -> None:
+def check_role_based_permissions(
+    invited_as: int, user_profile: UserProfile, *, require_atleast_admin: bool = False
```
`require_admin` would be a more traditional name for this parameter.

timabbott commented on zerver/views/invite.py:
```diff
@@ -29,13 +29,28 @@
+        ]
+        and not user_profile.is_realm_admin
+    ):
+        raise JsonableError(_("Must be an organization administrator"))
```
I think this block could be replaced by having the one caller pass `require_admin=invited_as in [PreregistrationUser.INVITE_AS["REALM_ADMIN"], ...]`, maybe with that variable being set on its own line, in the one caller that uses this.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,16 @@ format used by the Zulip server that they are interacting with.
+  can create invite link to realm.
+
+* `PATCH /realm`: Added a new parameter `create_multiuse_invite_group_id`
+  for changing the user_group whose members can create invite link.
```
The writing style here feels a lot more sloppy than existing API changelog entries; please read some similar ones and polish this.

timabbott commented on corporate/tests/test_stripe.py:
```diff
@@ -3662,7 +3667,18 @@ def create_realm(
+                    setattr(realm, setting_group_id_name, -1)
+                realm.save()
+                create_system_user_groups_for_realm(realm)
+                set_default_for_realm_permission_group_settings(realm)
```
I think a better change here would be to add a prep commit that makes this test call `do_create_realm` properly rather than trying to make a realm object hackily this other way.

timabbott commented on zerver/actions/realm_settings.py:
```diff
@@ -104,6 +105,43 @@ def do_set_realm_property(
 
 
+@transaction.atomic(durable=True)
+def do_set_realm_permission_group_settings(
```
`do_change_realm_permission_group_setting` (singular, and using `change`) feels clearer to me. We're changing a single group setting.

The reason I prefer "change" over "set" is that we're moving a pointer (the group ID), effectively, rather than changing a value.

timabbott commented on zerver/actions/realm_settings.py:
```diff
@@ -104,6 +105,43 @@ def do_set_realm_property(
+    user_group to update and and the user who initiated the update.
+    """
+
+    old_user_group_id = getattr(realm, name)
```
Let's change `name` to `setting_name`, and have an `assert setting_name in REALM_PERMISSION_GROUP_SETTING_ID_TO_GROUP_SETTING` line.

timabbott commented on zerver/actions/realm_settings.py:
```diff
@@ -104,6 +105,43 @@ def do_set_realm_property(
+        op="update",
+        property=name,
+        value=user_group.id,
+    )
```
As discussed in https://chat.zulip.org/#narrow/stream/378-api-design/topic/event.20for.20updating.20group.20setting/near/1600758, let's avoid the property/value format here.

timabbott commented on zerver/actions/realm_settings.py:
```diff
@@ -104,6 +105,43 @@ def do_set_realm_property(
+        value=user_group.id,
+    )
+
+    transaction.on_commit(lambda: send_event(realm, event, active_user_ids(realm.id)))
```
This is an old pattern; switch this to `send_event_on_commit`. @HemantUmre12 @sahil839 FYI too since you may have some PRs that need this tweak too.

timabbott commented on zerver/lib/events.py:
```diff
@@ -257,6 +257,9 @@ def fetch_initial_state_data(
         for property_name in Realm.property_types:
             state["realm_" + property_name] = getattr(realm, property_name)
 
+        for property_name in Realm.REALM_PERMISSION_GROUP_SETTING_ID_TO_GROUP_SETTING:
```
This could be accessing the shorter `REALM_PERMISSION_GROUP_SETTINGS`, right? I think there may be a few more cases in this PR too.

timabbott commented:
@Ujjawal3 let me know when you've gotten the feedback above resolved -- I think this is not too far from being mergeable once the above comments are resolved.

Ujjawal3 commented on zerver/views/realm.py:
```diff
@@ -269,7 +274,16 @@ def update_realm(
+        if k in realm.property_types:
+            req_vars[k] = v
+
+        for permissions_configuration in Realm.REALM_PERMISSION_GROUP_SETTINGS.values():
+            if k == permissions_configuration.id_field_name:
+                req_group_setting_vars[k] = v
```
I am not sure whether there is anything better which can be done here.

Ujjawal3 commented:
@timabbott I have addressed all the feedback. You can take a look now.

timabbott commented:
We might want to rework this to remove the `_id` suffix as proposed in #26243 -- I guess let's try to prioritize finishing that to confirm that theory, since if so, we'll save some change management if we merge this without the `_id` suffix in the API.

timabbott commented on corporate/tests/test_stripe.py:
```diff
@@ -3662,7 +3662,7 @@ def create_realm(
         ) -> Tuple[Realm, Optional[Customer], Optional[CustomerPlan], List[stripe.Invoice]]:
             realm_string_id = "realm_" + str(random.randrange(1, 1000000))
-            realm = Realm.objects.create(string_id=realm_string_id)
+            realm = do_create_realm(string_id=realm_string_id, name=realm_string_id, plan_type=1)
```
Please don't use numerical values like `1`; always use readable values like `Realm.PLAN_TYPE_FOO` instead.

sahil839 commented:
@Ujjawal3 I think you should update this since #26372 is merged now. Would good to get this merged as soon as Tim gets time to review this, since it will unblock a couple of other PRs as well. Feel free to ask any doubts to update this on CZO.

laurynmm commented:
@Ujjawal3 - Also ping me when it's updated and ready for an API documentation review.

Ujjawal3 commented on api_docs/changelog.md:
```diff
@@ -30,6 +30,12 @@ format used by the Zulip server that they are interacting with.
 * `PATCH /realm`: Added `create_multiuse_invite_group` parameter
   to support setting the user group whose members can create invite link.
 
+* `GET /invites`: Now, the endpoint response for non-admin users includes
+   the multiuse invites created by them.
+
+* `DEL /invites/multiuse/{invite_id}`: Allow non-admins users to revoke
+   multiuse invites created by them.
+
```
I am not sure whether we want a separate feature level entry for these API changes.

Ujjawal3 commented:
I have updated the PR with the following major changes:-
- Updated the API design to use the one used in #26243.
- Updated the code to  use new dropdown list widget.
- Added commits to allow non-admins to revoke multiuse invites created by them.
Also, I have added some comments at places where I think something can be done better.
@laurynmm this PR is ready for API documentation review.

sahil839 commented on zerver/actions/invites.py:
```diff
@@ -31,15 +31,15 @@
     event = dict(type="invites_changed")
-    admin_ids = [user.id for user in realm.get_admin_users_and_bots()]
-    send_event(realm, event, admin_ids)
+    send_event(realm, event, active_user_ids(realm.id))
```
Should we just sent the event to admins and the user who would be affected like the one who sent the invite only and not all the non-admin users?

Ujjawal3 replied:
Yes we can do that. So I think we should change the function to something like this
```
def notify_invites_changed(realm: Realm, acting_user: Optional[UserProfile]=None) -> None:
    event = dict(type="invites_changed")
    admin_ids = [user.id for user in realm.get_admin_users_and_bots()]
    recipient_ids=admin_ids
    if(acting_user):
        recipient_ids.append(acting_user.id)
   send_event(realm, event, recipient_ids)
```

sahil839 replied:
Yes, that seems correct.

sahil839 reviewed (commented):
Posted few comments.

laurynmm commented on help/invite-new-users.md:
```diff
@@ -57,8 +57,6 @@ permission to invite users.
 
 ## Create an invitation link
```
Maybe change to "Create a reusable invitation link" for consistency.

laurynmm reviewed (commented):
@Ujjawal3 - Thanks for pinging me for an API docs review and all the work you've done on this!

I had a few comments about the API documentation and I also looked at the help center documentation while I was at it.

Some general questions I have after reading through the documentation, but not the full code changes:
- Shouldn't there also be an changelog entry for the `POST /invites/multiuse` endpoint since non-admin users can now potentially use that endpoint? 
- I think we should land on a term to consistantly user for these links in both the API and help center documentation. I like "reusable invitation link", which is currently used in the help center docs. Also, this could be used in the UI for the organization setting panel, which could be good because it clarifies to the admin updating that setting that these aren't one-time use links.
- Question: could a user now have the ability to create a reusable invitation link, but not invite users via email? If so, we should clarify that in the help center documentation.
- Also, are there limitations for these reusable invitation links if I'm not an owner/admin? For example, I imagine that as a normal member that I wouldn't be able to invite someone to be an admin or moderator, correct? If so, that also should be clarified in the help center documentation updates.

Let me know if you have any questions about the points above or the other comments in my review. Thanks!

Ujjawal3 commented on zerver/tests/test_events.py:
```diff
@@ -2869,11 +2870,84 @@ def do_set_realm_property_test(self, name: str) -> None:
+                ),
+                state_change_expected=state_change_expected,
+                num_events=num_events,
+            )
```
I could not figure out how to define data type of events corresponding to group setting updates. We need to define required keys and optional keys but for these events there should be atleast one key from `Realm.REALM_GROUP_PERMISSIONS_SETTING` . I tried something like this but there was an error because of enum type.
```
    required_keys=[
      (EnumType([setting_name for setting_name in Realm.REALM_PERMISSION_GROUP_SETTINGS]),int),
    ]
```
I looked at other definitions but could not find any where we need to do such thing.

sahil839 replied:
I don't think you can use `EnumType` as a first element of a tuple.

I think what we can do is define something like `group_setting_update_data_type` as a `DictType` with only `optional_keys` and then check that atlease one field is present. Something similar to what we do for `check_user_group_update`.

Ujjawal3 commented on zerver/lib/import_realm.py:
```diff
@@ -950,12 +951,21 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
+    else:
+        for setting_name in Realm.REALM_PERMISSION_GROUP_SETTINGS:
+            if setting_name in data["zerver_realm"][0]:
+                del data["zerver_realm"][0][setting_name]
```
Yes, I have modified the test data and have removed this code.

Ujjawal3 commented:
> @Ujjawal3 - Thanks for pinging me for an API docs review and all the work you've done on this!
> 
> I had a few comments about the API documentation and I also looked at the help center documentation while I was at it.
> 
> Some general questions I have after reading through the documentation, but not the full code changes:
> 
> * Shouldn't there also be an changelog entry for the `POST /invites/multiuse` endpoint since non-admin users can now potentially use that endpoint?
> * I think we should land on a term to consistantly user for these links in both the API and help center documentation. I like "reusable invitation link", which is currently used in the help center docs. Also, this could be used in the UI for the organization setting panel, which could be good because it clarifies to the admin updating that setting that these aren't one-time use links.
> * Question: could a user now have the ability to create a reusable invitation link, but not invite users via email? If so, we should clarify that in the help center documentation.
> * Also, are there limitations for these reusable invitation links if I'm not an owner/admin? For example, I imagine that as a normal member that I wouldn't be able to invite someone to be an admin or moderator, correct? If so, that also should be clarified in the help center documentation updates.
> 
> Let me know if you have any questions about the points above or other comments in my review. Thanks!

@laurynmm thanks for the review. I have done the suggested changes.
Regarding the questions you asked
- Yeah, I missed the entry for `POST /invites/multiuse` endpoint earlier. I have added it now.
- Yes, now there are separate settings for both invitations method. So, it may be possible that a user can only create invite links or vice versa. I think the current help documentation clarifies it.
- Yes, there are limitations on the role which can be invited by a user.  Invitation for owners role can be only created by organisation owners and invitations for admins and moderators role can be only created  by administrators. We do have these same restrictions for inviting as through emails also. I do not think it is documented now for either method. I am not sure about the right place to document this for both methods.

Ujjawal3 commented:
@sahil839 I have addressed your comments also.

laurynmm commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,30 @@ format used by the Zulip server that they are interacting with.
+
+* `POST /invites/multiuse`: Non-admin users can now use this endpoint to
+  create reusable invite links. Previously only admins users could create
+  reusable invite links.
```
We'll want a comma after "Previously" when it starts a sentence. Again, it would be good to be consistently using "reusable invitation links". And let's modify the second sentence to focus on the change to the endpoint:

```markdown
* `POST /invites/multiuse`: Non-admin users can now use this endpoint
  to create reusable invitation links. Previously, this endpoint was
  restricted to admin users only.
```

laurynmm commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,30 @@ format used by the Zulip server that they are interacting with.
+  create reusable invite links. Previously only admins users could create
+  reusable invite links.
+
+* `GET /invites`:  Endpoint response for non-admin users now includes both
```
Small note, there's an extra space after the colon here. I always find it useful to look at the documentation changes in a browser to catch these little things.

laurynmm commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,30 @@ format used by the Zulip server that they are interacting with.
+
+* `DELETE /invites/multiuse/{invite_id}`: Non-admin users can now revoke reusable
+  invite links they have created. Previously, only admin users could create and
+  revoke reusable invite links.
```
Again, let's be consistent with "reusable invitation links" here.

laurynmm commented on help/restrict-account-creation.md:
```diff
@@ -51,17 +50,16 @@ Regardless of whether invitations are required, you can:
-while permission to send out individual email invitations is configurable, creating
-*reusable* invitation links is always limited to administrators.
+To protect your organization, permissions to send out individual email
+invitations and creating reusable invite links are configurable.
```
Again, I think we can just delete all the content before the header here, even the `{!owner-only.md!)` since that's at the top of this page.

laurynmm commented on zerver/openapi/zulip.yaml:
```diff
@@ -3940,6 +3940,19 @@ paths:
+                                        ID of the [user group](/api/get-user-groups) whose members are
+                                        allowed to create [reusable invitation
+                                        links](/help/invite-new-users#create-a-reusable-invitation-link)
+                                        to realm.
```
One more small tweak here and in the register response:

```markdown
The ID of the [user group](/api/get-user-groups) whose members are
allowed to create [reusable invitation
links](/help/invite-new-users#create-a-reusable-invitation-link)
to the organization.
```

laurynmm commented on web/templates/settings/organization_permissions_admin.hbs:
```diff
@@ -16,13 +16,18 @@
+                {{> ../dropdown_widget_with_label
+                  widget_name="realm_create_multiuse_invite_group"
+                  label=(t 'Who can create reusable invitation links')
+                  value_type="number"}}
```
I think this is being rendered a little differently than the other settings ...

<details>
<summary>screenshot as Iago</summary>

![Screenshot from 2023-08-10 15-41-14](https://github.com/zulip/zulip/assets/63245456/ec0db657-6adf-4126-aa87-d2bef86a3df5)
</details>

sahil839 replied:
Yeah we use the custom dropdown widget, like the ones used for `New stream announcements` setting for group-based settings. And other settings used the default `select` element.

sahil839 replied:
And the setting should be disabled for `Iago`. It is disabled for me.

Ujjawal3 replied:
Yeah, earlier the UI of dropdown widget was not consistent with the other select fields. I made changes in last push and now it looks consistent.

laurynmm replied:
Pulled up the latest changes and the color looks good now, but the down arrow/chevron in the new setting dropdown is still a bit bigger than the other ones (at least in Chrome and Firefox on Linux) ...

<details>
<summary>Organization settings - Joining the organization - Iago</summary>

![Screenshot from 2023-08-14 14-41-29](https://github.com/zulip/zulip/assets/63245456/c7ca91b6-ce49-4f8e-a52b-20478b9858dd)
</details>

timabbott replied:
This is a known issue with the component in general and we can deal with separately.

sahil839 commented on zerver/actions/invites.py:
```diff
@@ -36,10 +36,13 @@
+    recipient_ids = admin_ids
+    if acting_user:
+        recipient_ids.append(acting_user.id)
+    send_event(realm, event, recipient_ids)
```
I am not completely sure, but we might need a changelog entry for this too. @laurynmm FYI.

laurynmm replied:
Yeah, looking at https://zulip.com/api/get-events#invites_changed, we would need to update the event description for the new users that get these events as well as make a **Changes** note for the previous state (only admins received these events) and a corresponding entry/bullet point in the API changelog for feature level 202 (at this point).

sahil839 commented on zerver/tests/test_events.py:
```diff
@@ -923,6 +925,7 @@ def test_revoke_user_invite_event(self) -> None:
         events = self.verify_action(
             lambda: do_revoke_user_invite(prereg_users[0]),
             state_change_expected=False,
+            num_events=2,
```
I think duplicate events are being sent here for admins here.

Ujjawal3 replied:
Yes, earlier there were duplicate eevents. I have made changes now.

sahil839 commented on zerver/actions/invites.py:
```diff
@@ -431,7 +434,7 @@ def revoke_invites_generated_by_user(user_profile: UserProfile) -> None:
     Confirmation.objects.bulk_update(confirmations_to_revoke, ["expiry_date"])
     if len(confirmations_to_revoke):
-        notify_invites_changed(realm=user_profile.realm)
+        notify_invites_changed(realm=user_profile.realm, acting_user=user_profile)
```
I think we don't need to send the event here to `user_profile` since this is called as a result of deactivating the user.

Ujjawal3 commented:
I have documented the changes in events and have fixed the inconsistency in UI.

laurynmm commented on zerver/openapi/zulip.yaml:
```diff
@@ -960,10 +960,14 @@ paths:
-                                set of invitations changes; this tells clients they need to refetch
+                                A simple event sent when the set of invitations changes.
+                                This event is sent to organization administrators and creator of
+                                changed invitation; this tells clients they need to refetch
```
Needs to be "the creator of the changed invitation" because we're referring to a specific user and invitation. Might be nice to expand the first sentence for clarity too...

> A simple event sent when the set of invitations in the organization changes. This event is sent to organization administrators and the creator of the changed invitation; this tells clients ...

And in the **Changes** note below, maybe expand to "organization administrators" for consistency with the text above.

laurynmm commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,32 @@ format used by the Zulip server that they are interacting with.
+
+* [`GET /events`](/api/get-events): On changing invitations, invites changed event is
+  now sent to all admin users and the creator of changed invitation. Previously, this
+  event was only sent to admin users.  
```
Using feature level 163 as a template, how about ...

```markdown
* [`GET /events`](/api/get-events): When the set of invitations in an
  organization changes, an `invites_changed` event is now sent to the
  creator of the changed invitation, as well as all admin users.
  Previously, this event was only sent to admin users.
```

laurynmm reviewed (commented):
@Ujjawal3 - Thanks for updating the API doc changes! Two small comments on the `invites_changed` event documentation updates you've got here.

I didn't pull the PR down to QA the updates to the organization settings panel dropdown.

sahil839 commented on web/styles/settings.css:
```diff
@@ -748,6 +748,24 @@ input[type="checkbox"] {
+
+        & i {
+            font-size: 10px;
+            margin-right: -3px;
```
I think we can do this for other dropdown widgets as well. The widget for "Who can remove subscribers from this stream" setting is also different from other dropdowns on the page.

Would be good to have a discussion on CZO, but should not be a blocker for this PR.

timabbott replied:
Yeah, this would be good to discuss in chat.zulip.org @Ujjawal3 can you start such a thread and link back to it here?

Ujjawal3 replied:
I have started the discussion the discussion about this earlier [here](https://chat.zulip.org/#narrow/stream/101-design/topic/Large.20down.20arrow.20in.20dropdown.20list.20widget)

sahil839 reviewed (commented):
I don't have any more suggestions now. Looks good to me.

Ujjawal3 commented:
The last push was just for bumping feature level entry which I missed earlier. I have done the documentation changes.

Ujjawal3 commented:
@alya I have resolved the merge conflicts.

timabbott commented on zerver/models.py:
```diff
@@ -751,6 +759,17 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
+        create_multiuse_invite_group=GroupPermissionSetting(
+            require_system_group=True,
+            allow_internet_group=False,
+            allow_owners_group=False,
```
@sahil839 is there a reason `allow_owners_group` should be false here? I can see situations where one might want to limit it to just owners, but wasn't sure if there was some policy situation to think about here.

sahil839 replied:
I am not sure. I thought it was described as such in the issue, but it isn't. I don't think there would be any issues to allow `Owners only` option. @Ujjawal3 can you confirm if we are missing something here? If not then we can set `allow_owners_group` as true.

timabbott commented on web/src/server_events_dispatch.js:
```diff
@@ -246,6 +249,12 @@ export function dispatch_normal_event(event) {
 
+                        if (event.property === "invite_to_realm_policy") {
+                            settings_invites.update_invite_user_panel();
+                            ui_init.update_invite_user_option();
```
I'm pretty sure it's going to be an import cycle to have `server_events_disptach` call a function in the `ui_init` module. I think what we should do is move `update_invite_user_option` into either `gear_menu.js` to or maybe better a new `right_sidebar_ui.js` module that's responsible for that little component.

This is OK to merge since we're not trying to avoid import cycles in this module right now, but please start a #frontend discussion about this detail.

timabbott commented on web/templates/confirm_dialog/confirm_revoke_invite.hbs:
```diff
@@ -1,5 +1,9 @@
+{{#if referred_by}}
 <p>{{#tr}}Are you sure you want to revoke this invitation link created by <strong>{referred_by}</strong>?{{/tr}}</p>
 {{else}}
+<p>{{#tr}}Are you sure you want to revoke this invitation link ?{{/tr}}</p>
```
There shouldn't be a space before the `?` here; I'll just fix that before merging.

Also, I wonder if it'd be better to make the check "If the referrer is anyone other than me"?

timabbott commented on zerver/actions/invites.py:
```diff
@@ -36,10 +36,13 @@
-    send_event(realm, event, admin_ids)
+    recipient_ids = admin_ids
+    if acting_user and acting_user.id not in recipient_ids:
+        recipient_ids.append(acting_user.id)
```
This logic does not match the API documentation -- we're notifying the **acting user**, not the user who created the modified invitation. I'm pretty sure this implementation is thus wrong if say an administrator revokes an invitation that a normal user had created.

This is definitely a bug and we should fix it but it's also obscure enough that I'm not going to block merging this on doing so; we can just fix it in a near-term follow-up.

Ujjawal3 replied:
I think there is nothing wrong. We are passing the user who created the invitation not the one changing the invitation to `acting_user`. We must change the `acting_user` to some other name to avoid this confusion.

sahil839 replied:
Yeah, I think we can just rename the `acting_user` to something else. I was thinking of `referred_by` first, but that would also be confusing since we do not include which invite wass changed in the event. Probably something like `changed_invite_referrer` could be better.

timabbott replied:
Yeah `changed_invite_referrer` works for me.

timabbott commented on zerver/migrations/0470_set_default_value_for_create_multiuse_invite_group.py:
```diff
@@ -0,0 +1,57 @@
+                default_group = UserGroup.objects.get(
+                    name=default_group_name, realm=realm, is_system_group=True
+                )
+                realm.create_multiuse_invite_group = default_group
```
So I think we actually want the default for existing administrators that hadn't specified `NOBODY` as their policy to be `role:administrators`, because that way we don't adjust permissions for anyone upwards unexpectedly.

timabbott replied:
``` diff
diff --git a/zerver/migrations/0470_set_default_value_for_create_multiuse_invite_group.py b/zerver/migrations/0470_set_default_value_for_create_multiuse_invite_group.py
index f162fb3e05..6a82d6afea 100644
--- a/zerver/migrations/0470_set_default_value_for_create_multiuse_invite_group.py
+++ b/zerver/migrations/0470_set_default_value_for_create_multiuse_invite_group.py
@@ -11,34 +11,23 @@ def set_default_value_for_create_multiuse_invite_group(
     Realm = apps.get_model("zerver", "Realm")
     UserGroup = apps.get_model("zerver", "UserGroup")
 
-    Realm.POLICY_NOBODY = 6
-    Realm.POLICY_ADMINS_ONLY = 2
-
-    UserGroup.NOBODY_GROUP_NAME = "role:nobody"
     UserGroup.ADMINISTRATORS_GROUP_NAME = "role:administrators"
-    default_group_name = "role:moderators"
 
     for realm in Realm.objects.all():
-        if realm.create_multiuse_invite_group is None:
-            if realm.invite_to_realm_policy == Realm.POLICY_NOBODY:
-                nobody_group = UserGroup.objects.get(
-                    name=UserGroup.NOBODY_GROUP_NAME, realm=realm, is_system_group=True
-                )
-                realm.create_multiuse_invite_group = nobody_group
-
-            elif realm.invite_to_realm_policy == Realm.POLICY_ADMINS_ONLY:
-                admins_group = UserGroup.objects.get(
-                    name=UserGroup.ADMINISTRATORS_GROUP_NAME, realm=realm, is_system_group=True
-                )
-                realm.create_multiuse_invite_group = admins_group
-
-            else:
-                default_group = UserGroup.objects.get(
-                    name=default_group_name, realm=realm, is_system_group=True
-                )
-                realm.create_multiuse_invite_group = default_group
-
-            realm.save(update_fields=["create_multiuse_invite_group"])
+        if realm.create_multiuse_invite_group is not None:
+            continue
+
+        # Prior to the new create_multiuse_invite_group field being
+        # created, multi-use invitation links could only be created
+        # and managed by administrators, regardless of
+        # invite_to_realm_policy. We replicate that policy for the
+        # initial value of the new setting.
+        admins_group = UserGroup.objects.get(
+            name=UserGroup.ADMINISTRATORS_GROUP_NAME, realm=realm, is_system_group=True
+        )
+        realm.create_multiuse_invite_group = admins_group
+
+        realm.save(update_fields=["create_multiuse_invite_group"])
 
 
 class Migration(migrations.Migration):
```

timabbott commented on zerver/models.py:
```diff
@@ -751,6 +759,17 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
+            allow_internet_group=False,
+            allow_owners_group=False,
+            allow_nobody_group=True,
+            default_group_name=MODERATORS_GROUP_NAME,
```
After discussion with @alya, I think we should change this to `ADMINISTRATORS_GROUP_NAME`; we couldn't come up with a ton of use cases for moderator use of this, and I feel like that might better match user expectations in new organizations.

timabbott commented:
Thanks for all the work on this @Ujjawal3! I made a couple small changes to defaults, noted above, and pushed that version back here; the remaining comments I believe are all reasonable things to do as follow-ups:

* Most timely is https://github.com/zulip/zulip/issues/26652, which is a definitely bug in this logic that I expect nobody but us to notice, but we should fix while it's fresh.
* Discuss in #backend and then propose a better place to put the basic data about system groups (name, etc.): https://github.com/zulip/zulip/pull/25884/files#r1218280325
* Discuss in #frontend how to resolve https://github.com/zulip/zulip/pull/25884/files#r1317890034.
* Discuss in #frontend how to resolve https://github.com/zulip/zulip/pull/25884#discussion_r1294681898.
* Should `access_user_group_for_setting` accept a `permission_configuration` object rather than being passed a bunch of boolean values from that object? @sahil839 thoughts, but this seems to me like it might be a cleaner API.
* This is probably an "api design" conversation, but @sahil839 would it make sense to migrate `web/src/group_permission_settings.ts` to be something sent over the API by the server? The values there feel like things that end up being duplicated in the server and client code, and it seems like if the configuration were something that didn't need to be hardcoded on the client, that might be very handy. I think the main counter-argument is that we're not yet sure what set of properties we'll end up with long term... but I think we have enough of these now that it's worth thinking a bit as a group about what shape such an API might want to take.

@Ujjawal3 I know GSoC is ending; do you think you'll have time to take care of those follow-ups? If not, I expect @sahil839 can take over responsibility for them.

timabbott commented:
For `can_invite_others_to_realm`, I didn't like the name `can_send_email_invite_to_others` and changed it to `user_can_invite_users_by_email`, diff below. I think we may want to do a small follow-up to do the API change to rename the reference to `can_invite_others_to_realm` that's in the API; probably worth an "api design" discussion as I think it's possible a reasonable alternative would be to cut the field entirely; I'm not sure the web app is using it, nor are the mobile apps.

``` diff
diff --git a/web/src/admin.js b/web/src/admin.js
index 6cd1ea455a..a76ac4d0fe 100644
--- a/web/src/admin.js
+++ b/web/src/admin.js
@@ -160,7 +160,7 @@ export function build_page() {
         email_address_visibility_values: settings_config.email_address_visibility_values,
         waiting_period_threshold_dropdown_values:
             settings_config.waiting_period_threshold_dropdown_values,
-        can_send_email_invite_to_others: settings_data.user_can_send_email_invite_to_others(),
+        can_invite_users_by_email: settings_data.user_can_invite_users_by_email(),
         realm_invite_required: page_params.realm_invite_required,
         can_edit_user_groups: settings_data.user_can_edit_user_groups(),
         policy_values: settings_config.common_policy_values,
diff --git a/web/src/gear_menu.ts b/web/src/gear_menu.ts
index d28da78eaa..eb00adf2ee 100644
--- a/web/src/gear_menu.ts
+++ b/web/src/gear_menu.ts
@@ -126,7 +126,7 @@ export function initialize(): void {
         server_needs_upgrade: page_params.server_needs_upgrade,
         version_display_string: version_display_string(),
         apps_page_url: page_params.apps_page_url,
-        can_send_email_invite_to_others: settings_data.user_can_send_email_invite_to_others(),
+        can_invite_users_by_email: settings_data.user_can_invite_users_by_email(),
         corporate_enabled: page_params.corporate_enabled,
         is_guest: page_params.is_guest,
         login_link: page_params.development_environment ? "/devlogin/" : "/login/",
diff --git a/web/src/settings_data.ts b/web/src/settings_data.ts
index 6c2a7a8fa9..84de43f863 100644
--- a/web/src/settings_data.ts
+++ b/web/src/settings_data.ts
@@ -131,7 +131,7 @@ function user_has_permission(policy_value: number): boolean {
     return user_join_days >= page_params.realm_waiting_period_threshold;
 }
 
-export function user_can_send_email_invite_to_others(): boolean {
+export function user_can_invite_users_by_email(): boolean {
     if (
         page_params.realm_invite_to_realm_policy ===
         settings_config.email_invite_to_realm_policy_values.nobody.code
diff --git a/web/src/settings_invites.js b/web/src/settings_invites.js
index a3fcee9402..95bf102420 100644
--- a/web/src/settings_invites.js
+++ b/web/src/settings_invites.js
@@ -243,7 +243,7 @@ export function on_load_success(invites_data, initialize_event_handlers) {
 }
 
 export function update_invite_users_setting_tip() {
-    if (settings_data.user_can_send_email_invite_to_others() && !page_params.is_admin) {
+    if (settings_data.user_can_invite_users_by_email() && !page_params.is_admin) {
         $(".invite-user-settings-tip").hide();
         return;
     }
@@ -296,7 +296,7 @@ export function update_invite_users_setting_tip() {
 
 export function update_invite_user_panel() {
     update_invite_users_setting_tip();
-    if (!settings_data.user_can_send_email_invite_to_others()) {
+    if (!settings_data.user_can_invite_users_by_email()) {
         $("#admin-invites-list .invite-user-link").hide();
     } else {
         $("#admin-invites-list .invite-user-link").show();
diff --git a/web/src/ui_init.js b/web/src/ui_init.js
index 1bdc44c5af..99e27c009b 100644
--- a/web/src/ui_init.js
+++ b/web/src/ui_init.js
@@ -190,7 +190,7 @@ function initialize_left_sidebar() {
 }
 
 export function update_invite_user_option() {
-    if (!settings_data.user_can_send_email_invite_to_others()) {
+    if (!settings_data.user_can_invite_users_by_email()) {
         $("#right-sidebar .invite-user-link").hide();
     } else {
         $("#right-sidebar .invite-user-link").show();
diff --git a/web/templates/gear_menu.hbs b/web/templates/gear_menu.hbs
index df6a45bb16..64d51fcbce 100644
--- a/web/templates/gear_menu.hbs
+++ b/web/templates/gear_menu.hbs
@@ -155,7 +155,7 @@
             </li>
             {{/if}}
             <li class="divider hidden-for-spectators" role="presentation"></li>
-            {{#if can_send_email_invite_to_others}}
+            {{#if can_invite_users_by_email}}
             <li role="presentation">
                 <a class="invite-user-link" role="menuitem">
                     <i class="fa fa-user-plus" aria-hidden="true"></i> {{t 'Invite users' }}
diff --git a/web/tests/settings_data.test.js b/web/tests/settings_data.test.js
index 469815b89d..5290f43129 100644
--- a/web/tests/settings_data.test.js
+++ b/web/tests/settings_data.test.js
@@ -158,7 +158,7 @@ test_policy(
 test_policy(
     "user_can_invite_others_to_realm",
     "realm_invite_to_realm_policy",
-    settings_data.user_can_send_email_invite_to_others,
+    settings_data.user_can_invite_users_by_email,
 );
 test_policy(
     "user_can_move_messages_between_streams",
@@ -269,7 +269,7 @@ run_test("user_can_invite_others_to_realm_nobody_case", () => {
     page_params.is_guest = false;
     page_params.realm_invite_to_realm_policy =
         settings_config.email_invite_to_realm_policy_values.nobody.code;
-    assert.equal(settings_data.user_can_send_email_invite_to_others(), false);
+    assert.equal(settings_data.user_can_invite_users_by_email(), false);
 });
 
 run_test("user_can_create_web_public_streams", () => {
diff --git a/zerver/lib/events.py b/zerver/lib/events.py
index 53291d7531..54a0a7eee7 100644
--- a/zerver/lib/events.py
+++ b/zerver/lib/events.py
@@ -492,7 +492,7 @@ def fetch_initial_state_data(
         )
         state["can_create_web_public_streams"] = settings_user.can_create_web_public_streams()
         state["can_subscribe_other_users"] = settings_user.can_subscribe_other_users()
-        state["can_invite_others_to_realm"] = settings_user.can_send_email_invite_to_others()
+        state["can_invite_others_to_realm"] = settings_user.can_invite_users_by_email()
         state["is_admin"] = settings_user.is_realm_admin
         state["is_owner"] = settings_user.is_realm_owner
         state["is_moderator"] = settings_user.is_moderator
@@ -894,7 +894,7 @@ def apply_event(
                     state["can_subscribe_other_users"] = user_profile.can_subscribe_other_users()
                     state[
                         "can_invite_others_to_realm"
-                    ] = user_profile.can_send_email_invite_to_others()
+                    ] = user_profile.can_invite_users_by_email()
 
                     if state["is_guest"]:
                         state["realm_default_streams"] = []
diff --git a/zerver/models.py b/zerver/models.py
index 3b44ddb838..e356d2c695 100644
--- a/zerver/models.py
+++ b/zerver/models.py
@@ -2177,7 +2177,7 @@ class UserProfile(AbstractBaseUser, PermissionsMixin, UserBaseSettings):  # type
     def can_subscribe_other_users(self) -> bool:
         return self.has_permission("invite_to_stream_policy")
 
-    def can_send_email_invite_to_others(self) -> bool:
+    def can_invite_users_by_email(self) -> bool:
         return self.has_permission("invite_to_realm_policy")
 
     def can_create_multiuse_invite_to_realm(self) -> bool:
diff --git a/zerver/tests/test_invite.py b/zerver/tests/test_invite.py
index ed4b929364..e5f41de230 100644
--- a/zerver/tests/test_invite.py
+++ b/zerver/tests/test_invite.py
@@ -801,7 +801,7 @@ class InviteUserTest(InviteUserBase):
     def test_can_invite_others_to_realm(self) -> None:
         def validation_func(user_profile: UserProfile) -> bool:
             user_profile.refresh_from_db()
-            return user_profile.can_send_email_invite_to_others()
+            return user_profile.can_invite_users_by_email()
 
         realm = get_realm("zulip")
         do_set_realm_property(
diff --git a/zerver/views/invite.py b/zerver/views/invite.py
index 8b032a551c..ae4a7d30de 100644
--- a/zerver/views/invite.py
+++ b/zerver/views/invite.py
@@ -59,7 +59,7 @@ def invite_users_backend(
     ),
     stream_ids: List[int] = REQ(json_validator=check_list(check_int)),
 ) -> HttpResponse:
-    if not user_profile.can_send_email_invite_to_others():
+    if not user_profile.can_invite_users_by_email():
         # Guest users case will not be handled here as it will
         # be handled by the decorator above.
         raise JsonableError(_("Insufficient permission"))
```

timabbott commented:
So I was more or less ready to merge this, but I noticed that we currently allow "guests" as an option for this new setting, which we don't allow for the similar "who can do an invite link" setting, so probably is a bad idea to support.

I'm going to try adding an `allow_everyone_group` property to address this as a new final commit.

timabbott commented:
OK, added a commit to that effect; @sahil839 I'd greatly appreciate your doing a review on that and any follow-ups; I'm pretty sure that it probably makes some of the `zulip.yaml` notes about which values are allowed slightly inaccurate; that's probably OK to fix in a follow-up so that we can get this integrated and build additional things on top of it. 

I'm now pretty convinced that we want to be passing the `permissions_config` object around.

Ujjawal3 commented:
> 

I can work on the follow-ups after 15. I have exams next week so currently busy preparing for it.

Ujjawal3 commented:
> Thanks for all the work on this @Ujjawal3! I made a couple small changes to defaults, noted above, and pushed that version back here; the remaining comments I believe are all reasonable things to do as follow-ups:
> 
> * Most timely is [Fix live update when an administrator revokes an invitation created by another user #26652](https://github.com/zulip/zulip/issues/26652), which is a definitely bug in this logic that I expect nobody but us to notice, but we should fix while it's fresh.
> * Discuss in #backend and then propose a better place to put the basic data about system groups (name, etc.): https://github.com/zulip/zulip/pull/25884/files#r1218280325
> * Discuss in #frontend how to resolve https://github.com/zulip/zulip/pull/25884/files#r1317890034.
> * Discuss in #frontend how to resolve [invitations: Add a new setting `Who can create multiuse invite link`. #25884 (comment)](https://github.com/zulip/zulip/pull/25884#discussion_r1294681898).
> * Should `access_user_group_for_setting` accept a `permission_configuration` object rather than being passed a bunch of boolean values from that object? @sahil839 thoughts, but this seems to me like it might be a cleaner API.
> * This is probably an "api design" conversation, but @sahil839 would it make sense to migrate `web/src/group_permission_settings.ts` to be something sent over the API by the server? The values there feel like things that end up being duplicated in the server and client code, and it seems like if the configuration were something that didn't need to be hardcoded on the client, that might be very handy. I think the main counter-argument is that we're not yet sure what set of properties we'll end up with long term... but I think we have enough of these now that it's worth thinking a bit as a group about what shape such an API might want to take.
> 
> @Ujjawal3 I know GSoC is ending; do you think you'll have time to take care of those follow-ups? If not, I expect @sahil839 can take over responsibility for them.

I can work on follow-ups after 15. I have exams from next week so currently busy in preparing for it.

sahil839 commented:
I had a look at the commit for adding `allow_everyone_group` and it looks good. I can do the required API docs update if there are any and push to this PR, but just wanted to confirm whether we want to fix the variable in `notify_invites_changed` as [discussed here](https://github.com/zulip/zulip/pull/25884#discussion_r1318178236) now itself since the PR was not merged due to lint issues.

timabbott commented:
> And another question, whether we would sent different objects for realm-level, stream-level and group-level settings or just a single object with all settings

Probably the right idea is 3 objects with the same type. Not sure about naming. 

I think `group_permission_settings.ts` can still be there to extend the server data if we feel the need to do that. But my main concern is having the pretty long list of settings get out of sync with each other.

In any case, this is now merged after @sahil839 fixed the little CI issues and https://github.com/zulip/zulip/pull/25884#discussion_r1318884468; huge thanks for all the work getting us here @Ujjawal3! It's gonna be like 5x easier to do future realm settings with a group type now that we've done this groundwork.

Ujjawal3 commented:
> Thanks for all the work on this @Ujjawal3! I made a couple small changes to defaults, noted above, and pushed that version back here; the remaining comments I believe are all reasonable things to do as follow-ups:
> 
> * Most timely is [Fix live update when an administrator revokes an invitation created by another user #26652](https://github.com/zulip/zulip/issues/26652), which is a definitely bug in this logic that I expect nobody but us to notice, but we should fix while it's fresh.
> * Discuss in #backend and then propose a better place to put the basic data about system groups (name, etc.): https://github.com/zulip/zulip/pull/25884/files#r1218280325
> * Discuss in #frontend how to resolve https://github.com/zulip/zulip/pull/25884/files#r1317890034.
> * Discuss in #frontend how to resolve [invitations: Add a new setting `Who can create multiuse invite link`. #25884 (comment)](https://github.com/zulip/zulip/pull/25884#discussion_r1294681898).
> * Should `access_user_group_for_setting` accept a `permission_configuration` object rather than being passed a bunch of boolean values from that object? @sahil839 thoughts, but this seems to me like it might be a cleaner API.
> * This is probably an "api design" conversation, but @sahil839 would it make sense to migrate `web/src/group_permission_settings.ts` to be something sent over the API by the server? The values there feel like things that end up being duplicated in the server and client code, and it seems like if the configuration were something that didn't need to be hardcoded on the client, that might be very handy. I think the main counter-argument is that we're not yet sure what set of properties we'll end up with long term... but I think we have enough of these now that it's worth thinking a bit as a group about what shape such an API might want to take.
> 
> @Ujjawal3 I know GSoC is ending; do you think you'll have time to take care of those follow-ups? If not, I expect @sahil839 can take over responsibility for them.

@sahil839 can you work on these follow-ups? Reading the work it looks like it will take a while to do them Tim has asked me to resolve the merge conflicts in #24580 and it has some significant merge conflicts, so currently it would not be possible for me to work on this.

sahil839 commented:
Ok, I can work on these as required for my further work on #10970 and then handle the remaining if any after that.
